File: 948-bag-of-tokens/948-bag-of-tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        dq = deque(sorted(tokens))
        res = count = 0
        while dq and (dq[0] <= power or count):
            if dq[0] <= power:
                power -= dq.popleft()
                count += 1
            elif count:
                power += dq.pop()
                count -= 1
            res = max(res, count)
        return res

File: 948-bag-of-tokens/948-bag-of-tokens.js
/**
 * @param {number[]} tokens
 * @param {number} power
 * @return {number}
 */
var bagOfTokensScore = function(tokens, power) {
  tokens.sort((a, b) => a - b);
  let [res, point, l, r] = [0, 0, 0, tokens.length - 1];
  
  while (l <= r) {
    if (tokens[l] <= power) {
      power -= tokens[l++];
      res = Math.max(res, ++point);
    }
    else if (point > 0) {
      point--;
      power += tokens[r--];
    }
    else break;
  }
  
  return res;
};